<a href="https://colab.research.google.com/github/svyatoslavna/ml_hw/blob/main/trees_hw.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание: cравнение древовидных алгоритмов на NLP-данных

**Задача:** oбучить разные древовидные модели на текстах новостей, замерить test accuracy и время обучения, заполнить таблицы.

## 1. Загрузка и подготовка данных (код дан)

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# Загружаем 3 категории
categories = ['rec.sport.baseball', 'sci.space', 'comp.graphics']
newsgroups = fetch_20newsgroups(subset='all', categories=categories, shuffle=True, random_state=42)

# TF-IDF векторизация
vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(newsgroups.data)
y = newsgroups.target

# Разделение на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки: {X_train.shape}")
print(f"Размер тестовой выборки: {X_test.shape}")

Размер обучающей выборки: (2363, 5000)
Размер тестовой выборки: (591, 5000)


## 2. Образец: Дерево решений (Decision Tree)

**Гиперпараметр:** `max_depth`

In [13]:
from sklearn.tree import DecisionTreeClassifier

results_tree = []

# Образец: мы делаем перебор значений max_depth = [3, 5, 10, 20, None])
# Для каждого: замер времени, обучение, accuracy на тесте, сохранение в results_tree

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = DecisionTreeClassifier(max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_tree.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"max_depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_tree = pd.DataFrame(results_tree)
print("\nТаблица 1. Дерево решений")
print(df_tree.to_markdown(index=False))

max_depth=3: acc=0.6244, time=0.35s
max_depth=5: acc=0.6971, time=0.57s
max_depth=10: acc=0.7716, time=0.68s
max_depth=20: acc=0.8409, time=0.70s
max_depth=None: acc=0.8697, time=0.92s

Таблица 1. Дерево решений
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.35 |
|           5 |          0.6971 |       0.57 |
|          10 |          0.7716 |       0.68 |
|          20 |          0.8409 |       0.7  |
|         nan |          0.8697 |       0.92 |


## 3. Заполните таблицы для остальных алгоритмов

По аналогии с образцом обучите следующие модели и запишите результаты.

### 3.1. Случайный лес (Random Forest)

**Гиперпараметр:** `max_depth` (фиксируем `n_estimators=100`)

| max_depth | test_accuracy | time_sec |
|-----------|---------------|----------|
| 5         |               |          |
| 10        |               |          |
| 20        |               |          |
| None      |               |          |

In [12]:
from sklearn.ensemble import RandomForestClassifier

results_rf = []

for depth in [5, 10, 20, None]:
    start = time.time()
    clf = RandomForestClassifier(
        max_depth=depth,
        n_estimators=100,
        random_state=42
        )
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_rf.append({
        'max_depth': depth,
        'test_accuracy': round(acc, 4),
        'time_sec': round(fit_time, 2)
        })
    print(f"max_depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_rf = pd.DataFrame(results_rf)
print("\nТаблица 2. Случайный лес")
print(df_rf.to_markdown(index=False))

max_depth=5: acc=0.9002, time=0.62s
max_depth=10: acc=0.9154, time=1.05s
max_depth=20: acc=0.9357, time=1.55s
max_depth=None: acc=0.9526, time=3.61s

Таблица 2. Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           5 |          0.9002 |       0.62 |
|          10 |          0.9154 |       1.05 |
|          20 |          0.9357 |       1.55 |
|         nan |          0.9526 |       3.61 |


### 3.2. Градиентный бустинг (Gradient Boosting)

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100, max_depth=3`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.01          |               |          |
| 0.05          |               |          |
| 0.1           |               |          |
| 0.5           |               |          |

In [14]:
from sklearn.ensemble import GradientBoostingClassifier

results_gb = []

for learning_rate in [0.01, 0.05, 0.1, 0.5]:
    start = time.time()
    clf = GradientBoostingClassifier(
        learning_rate=learning_rate,
        max_depth=3,
        n_estimators=100,
        random_state=42
        )
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_gb.append({
        'learning_rate': learning_rate,
        'test_accuracy': round(acc, 4),
        'time_sec': round(fit_time, 2)
        })
    print(f"learning_rate={learning_rate}: acc={acc:.4f}, time={fit_time:.2f}s")

df_gb = pd.DataFrame(results_gb)
print("\nТаблица 3. Градиентный бустинг")
print(df_gb.to_markdown(index=False))

learning_rate=0.01: acc=0.8782, time=42.36s
learning_rate=0.05: acc=0.9475, time=57.02s
learning_rate=0.1: acc=0.9577, time=62.68s
learning_rate=0.5: acc=0.9662, time=61.66s

Таблица 3. Градиентный бустинг
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.01 |          0.8782 |      42.36 |
|            0.05 |          0.9475 |      57.02 |
|            0.1  |          0.9577 |      62.68 |
|            0.5  |          0.9662 |      61.66 |


### 3.3. XGBoost

**Гиперпараметр:** `gamma` (фиксируем `n_estimators=100, learning_rate=0.1`)

| gamma | test_accuracy | time_sec |
|-------|---------------|----------|
| 0     |               |          |
| 0.1   |               |          |
| 0.5   |               |          |
| 1.0   |               |          |

In [16]:
from xgboost import XGBClassifier

results_xgb = []

for gamma in [0, 0.1, 0.5, 1]:
    start = time.time()
    clf = XGBClassifier(
        gamma=gamma,
        n_estimators=100,
        learning_rate=0.1,
        eval_metric='logloss',
        random_state=42
        )
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_xgb.append({'gamma': gamma, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"gamma={gamma}: acc={acc:.4f}, time={fit_time:.2f}s")


df_xgb = pd.DataFrame(results_xgb)
print("\nТаблица 4. XGBoost")
print(df_xgb.to_markdown(index=False))


# в комментарии было дано еще одно условие:
# установить use_label_encoder=False;
# но вылезло такое предупреждение:

# WARNING: /__w/xgboost/xgboost/src/learner.cc:782:
# Parameters: { "use_label_encoder" } are not used.

# и в документации не нашла этот параметр почему-то...
# поэтому эту нерабочую часть кода я удалила

gamma=0: acc=0.9459, time=34.00s
gamma=0.1: acc=0.9442, time=35.05s
gamma=0.5: acc=0.9492, time=31.69s
gamma=1: acc=0.9492, time=31.61s

Таблица 4. XGBoost
|   gamma |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|     0   |          0.9459 |      34    |
|     0.1 |          0.9442 |      35.05 |
|     0.5 |          0.9492 |      31.69 |
|     1   |          0.9492 |      31.61 |


### 3.4. AdaBoost

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.5           |               |          |
| 1.0           |               |          |
| 1.5           |               |          |
| 2.0           |               |          |

In [17]:
from sklearn.ensemble import AdaBoostClassifier

results_ada = []

for learning_rate in [0.5, 1, 1.5, 2]:
    start = time.time()
    clf = AdaBoostClassifier(
        learning_rate=learning_rate,
        n_estimators=100,
        random_state=42
        )
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_ada.append({
        'learning_rate': learning_rate,
        'test_accuracy': round(acc, 4),
        'time_sec': round(fit_time, 2)
        })
    print(f"learning_rate={learning_rate}: acc={acc:.4f}, time={fit_time:.2f}s")

df_ada = pd.DataFrame(results_ada)
print("\nТаблица 5. AdaBoost")
print(df_ada.to_markdown(index=False))

learning_rate=0.5: acc=0.8629, time=7.79s
learning_rate=1: acc=0.9205, time=5.39s
learning_rate=1.5: acc=0.9255, time=5.98s
learning_rate=2: acc=0.8731, time=5.64s

Таблица 5. AdaBoost
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |       7.79 |
|             1   |          0.9205 |       5.39 |
|             1.5 |          0.9255 |       5.98 |
|             2   |          0.8731 |       5.64 |


### 3.5. LightGBM

**Гиперпараметр:** `num_leaves` (фиксируем `n_estimators=100, learning_rate=0.1`)

| num_leaves | test_accuracy | time_sec |
|------------|---------------|----------|
| 15         |               |          |
| 31         |               |          |
| 63         |               |          |
| 127        |               |          |

In [22]:
from lightgbm import LGBMClassifier

results_lgbm = []

for num_leaves in [15, 31, 63, 127]:
    start = time.time()
    clf = LGBMClassifier(
        num_leaves=num_leaves,
        n_estimators=100,
        learning_rate=0.1,
        n_jobs=-1,
        verbosity=-1,      # добавила, чтоб помолчал:) а то принтил много всего
        random_state=42
        )
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_lgbm.append({
        'num_leaves': num_leaves,
        'test_accuracy': round(acc, 4),
        'time_sec': round(fit_time, 2)
        })
    print(f"num_leaves={num_leaves}: acc={acc:.4f}, time={fit_time:.2f}s")

df_lgbm = pd.DataFrame(results_lgbm)
print("\nТаблица 6. LightGBM")
print(df_lgbm.to_markdown(index=False))

num_leaves=15: acc=0.9662, time=14.76s
num_leaves=31: acc=0.9594, time=13.35s
num_leaves=63: acc=0.9577, time=30.47s
num_leaves=127: acc=0.9594, time=21.23s

Таблица 6. LightGBM
|   num_leaves |   test_accuracy |   time_sec |
|-------------:|----------------:|-----------:|
|           15 |          0.9662 |      14.76 |
|           31 |          0.9594 |      13.35 |
|           63 |          0.9577 |      30.47 |
|          127 |          0.9594 |      21.23 |


### 3.6. CatBoost

**Гиперпараметр:** `depth` (фиксируем `iterations=100, learning_rate=0.1`)

| depth | test_accuracy | time_sec |
|-------|---------------|----------|
| 3     |               |          |
| 5     |               |          |
| 7     |               |          |
| 10    |               |          |

In [25]:
!pip install catboost -q

In [26]:
from catboost import CatBoostClassifier

results_cat = []

for depth in [3, 5, 7, 10]:
    start = time.time()
    clf = CatBoostClassifier(
        depth=depth,
        verbose=0,
        iterations=100,
        learning_rate=0.1,
        random_state=42
        )
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_cat.append({
        'depth': depth,
        'test_accuracy': round(acc, 4),
        'time_sec': round(fit_time, 2)
        })
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_cat = pd.DataFrame(results_cat)
print("\nТаблица 7. CatBoost")
print(df_cat.to_markdown(index=False))

depth=3: acc=0.9340, time=27.61s
depth=5: acc=0.9425, time=40.25s
depth=7: acc=0.9442, time=113.26s
depth=10: acc=0.9526, time=888.10s

Таблица 7. CatBoost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       3 |          0.934  |      27.61 |
|       5 |          0.9425 |      40.25 |
|       7 |          0.9442 |     113.26 |
|      10 |          0.9526 |     888.1  |


## 4. Итоговая таблица лучших результатов

In [27]:
# Соберите лучшие результаты из каждой таблицы вручную или автоматически

models_map = {
    'Decision Tree': (df_tree, 'max_depth'),
    'Random Forest': (df_rf, 'max_depth'),
    'Gradient Boosting': (df_gb, 'learning_rate'),
    'XGBoost': (df_xgb, 'gamma'),
    'AdaBoost': (df_ada, 'learning_rate'),
    'LightGBM': (df_lgbm, 'num_leaves'),
    'CatBoostClassifier': (df_cat, 'depth')
}

summary= []

for model_name, (df, param) in models_map.items():
    best_idx = df['test_accuracy'].idxmax()
    best_params = f"{param}={df.loc[best_idx, param]}"

    summary.append({
        'Algorithm': model_name,
        'Best params': best_params,
        'Best test acc': df.loc[best_idx, 'test_accuracy'],
        'Time (sec)': df.loc[best_idx, 'time_sec']
    })

summary = pd.DataFrame(summary)

print("Итоговая таблица лучших результатов")
print(summary.to_markdown(index=False))

Итоговая таблица лучших результатов
| Algorithm          | Best params       |   Best test acc |   Time (sec) |
|:-------------------|:------------------|----------------:|-------------:|
| Decision Tree      | max_depth=nan     |          0.8697 |         0.92 |
| Random Forest      | max_depth=nan     |          0.9526 |         3.61 |
| Gradient Boosting  | learning_rate=0.5 |          0.9662 |        61.66 |
| XGBoost            | gamma=0.5         |          0.9492 |        31.69 |
| AdaBoost           | learning_rate=1.5 |          0.9255 |         5.98 |
| LightGBM           | num_leaves=15     |          0.9662 |        14.76 |
| CatBoostClassifier | depth=10          |          0.9526 |       888.1  |


In [29]:
fastest = summary.loc[summary['Time (sec)'].idxmin()]
slowest = summary.loc[summary['Time (sec)'].idxmax()]

# во раз сколько самый быстрый быстрее самого медленного
speed_ratio = slowest['Time (sec)'] / fastest['Time (sec)']
speed_ratio

np.float64(965.3260869565217)

## 5. Вопросы (ответить текстом в следующей ячейке)

1. Какой алгоритм показал максимальную точность? Какие параметры к этому привели?

2. Какой алгоритм быстрее всего обучался? Во сколько раз он быстрее самого медленного?

3. У каких алгоритмов наблюдалось переобучение? При каких параметрах?

4. Какой алгоритм вы выбрали бы для продакшн-системы с миллионом текстов? Почему?

# Напишите свои ответы здесь

**Ответ 1:** Таких алгоритмов два.
- Градиентный бустинг

|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.5  |          0.9662 |      61.66 |


- LightGBM

|   num_leaves |   test_accuracy |   time_sec |
|-------------:|----------------:|-----------:|
|           15 |          0.9662 |      14.76

_____
**Ответ 2:**
*Если речь о лучших алгоритмах, т.е. тех, что в таблице summary, то...*

Самый быстрый:
- Дерево решений

|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|         nan |          0.8697 |       0.92 |

Самый медленный:

- CatBoost

|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|      10 |          0.9526 |     888.1  |

Самый быстрый (из лучших) алгоритм быстрее самого медленного в 965 раз.

_______
**Ответ 3:**

Переобучение наблюдалось у AdaBoost (на learning_rate = 2).

- AdaBoost

|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |       7.79 |
|             1   |          0.9205 |       5.39 |
|             1.5 |          0.9255 |       5.98 |
|         **2**   |      **0.8731** |       **5.64** |

Это понятно по тому, что с ростом learning_rate до 2 резко упала Accuracy до уровня, близкого модели с минимальным learning_rate = 0.5.
_________

**Ответ 4:**

*Если выбирать из тех моделей, что записаны в лучшие,* то для продакшн-системы с миллионом текстов я бы выбрала следующий алгоритм:

- Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|         nan |          0.9526 |       3.61 |

Для нас важны два показателя:
  - Качество (высокая Accuracy)
  - Скорость (т.к. очень много текстов)

В этом алгоритме сочетаются высокая точность и большая скорость относительно других моделей. Наиболее быстрый алгоритм - Дерево решений - сильно отстаёт по качеству (0.8697) от Случайного леса, при этом все остальные модели, которые обгоняют Случайный лес по качеству, обгоняют его не более, чем на две сотые, при этом они в разы медленнее.

*Если же мы посмотрим на все-все модели*, в том числе на те, что **не** были записаны в лучшие, мы обнаружим, что лучшие модели иногда по Accuracy не всегда принципиально отличаются от остальных, при этом времени на работу этих чуть менее точных алгоритмов часто требуется сильно меньше. Сравним все RandomForest модели:

- Случайный лес

|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           5 |          0.9002 |       0.62 |
|          10 |          0.9154 |       1.05 |
|      **20** |      **0.9357** |   **1.55** |
|         nan |          0.9526 |       3.61 |

В итоге моих размышлений я пришла к выводу, что лучше всего взять Случайный лес с max_depth = 20, т.к., опять же, Accuracy от модели с max_depth = None отличается не сильно (менее, чем на две сотые), но скорость модели с max_depth = 20 больше в 2,3 раза.

Выбор очевиден:

- Случайный лес

|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|      **20** |      **0.9357** |   **1.55** |

## Критерии оценки

| Что оценивается | Баллы |
|----------------|-------|
| Заполнены все 7 таблиц (правильно собраны accuracy и время) | 3 |
| Заполнена итоговая таблица лучших результатов | 1 |
| Ответы на 4 вопроса (по 0.5 балла) | 2 |
| Код воспроизводим (random_state=42, порядок ячеек корректен) | 1 |
| **Качество кода** (отсутствие дублирования, осмысленные имена переменных, циклы вместо копипасты, использование .to_markdown()) | 3 |
| **Итого** | **10** |

### Детали по качеству кода (+3 балла):
- **+1** — использование единого шаблона для всех экспериментов (цикл + list.append + pd.DataFrame)
- **+1** — правильное форматирование вывода (округление времени и accuracy до 2-4 знаков)
- **+1** — читаемые имена переменных, комментарии, отсутствие `eval()`, `exec()` и магических чисел

**Штрафы:**
- -1 балл, если код не запускается без ошибок
- -1 балл, если отсутствует `random_state=42` в любой из моделей
- -1 балл, если таблицы выведены криво (не через `to_markdown` или нечитаемый `print`)